src/rust/engine/process_execution/src/children.rs

```diff
@@ -47,14 +51,62 @@ impl ManagedChild {
     })
   }
 
-  /// Kill the process's unique PGID or return an error if we don't have a PID or cannot kill.
-  pub fn kill_pgid(&mut self) -> Result<(), String> {
+  fn get_pgid(&self) -> Result<Pid, String> {
     let pid = self.id().ok_or_else(|| "Process had no PID.".to_owned())?;
     let pgid = getpgid(Some(Pid::from_raw(pid as i32)))
       .map_err(|e| format!("Could not get process group id of child process: {}", e))?;
-    // Kill the negative PGID to kill the entire process group.
-    signal::kill(Pid::from_raw(-pgid.as_raw()), signal::Signal::SIGKILL)
+    Ok(pgid)
+  }
+
+  /// Send a signal to the child process group.
+  fn signal_pg<T: Into<Option<signal::Signal>>>(&mut self, signal: T) -> Result<(), String> {
+    let pgid = self.get_pgid()?;
+    // the negative PGID will signal the entire process group.
+    signal::kill(Pid::from_raw(-pgid.as_raw()), signal)
       .map_err(|e| format!("Failed to interrupt child process group: {}", e))?;
+    Ok(())
+  }
+
+  fn has_child_exited(&mut self) -> bool {
+    match self.child.try_wait() {
+      Ok(Some(_)) => true,
+      Ok(None) => false,
+      Err(_) => false,
+    }
+  }
+
+  /// Attempt to gracefully shutdown the process.
+  ///
+  /// This will send a SIGINT to the process and give it a chance to shutdown gracefully. If the
+  /// process does not respond to the SIGINT within a fixed interval, a SIGKILL will be sent.
+  ///
+  /// This method *will* block the current thread but will do so for a bounded amount of time.
+  pub fn graceful_shutdown(&mut self) -> Result<(), String> {
+    // send sigint
+    self.signal_pg(signal::Signal::SIGINT)?;
+
+    let deadline = time::Instant::now() + GRACEFUL_SHUTDOWN_WAIT_TIME;
+    loop {
+      if self.has_child_exited() {
+        // process terminated gracefully, mark this as killed to avoid unecessary additional cleanup
+        self.killed.store(true, Ordering::SeqCst);
+        break;
+      }
+
+      if time::Instant::now() >= deadline {
+        // we passed the deadline and process still has not shut down, so kill it
+        self.kill_pgid()?;
+        break;
+      }
+
+      thread::sleep(GRACEFUL_SHUTDOWN_POLL_TIME);
+    }
+    Ok(())
+  }
+
+  /// Kill the process's unique PGID or return an error if we don't have a PID or cannot kill.
+  pub fn kill_pgid(&mut self) -> Result<(), String> {
```

Comment:
```suggestion
  /// Kill the process's unique PGID or return an error if we don't have a PID or cannot kill.
  fn kill_pgid(&mut self) -> Result<(), String> {
```
Only used internally now, it looks like?